static/app/components/slot/slot.tsx

```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+
+interface SlotProps extends React.ComponentProps<any> {
+  children?: React.ReactNode;
+}
+
+/**
+ * The slot Component is used internally inside the design system to implement the asChild slot pattern.
+ * @internal
+ */
+export function Slot({children, ...props}: SlotProps): React.ReactNode {
+  // Destructure child, or else we'll just end up rendering it as the child prop
+  // and negate the entire slot functionality.
+  if (React.isValidElement(children)) {
+    return React.cloneElement(children, mergeProps(props, children.props));
+  }
+
+  // React.Children.only will throw if there are multiple children.
+  return process.env.NODE_ENV === 'production' ? null : React.Children.only(children);
+}
```

Comment:
could we get there by making the `children` prop typed better? Like, instead of `React.ReactNode`, we’d do `React.ReactElement`.

This wouldn’t allow multiple children, and it would also take away invalid children like `string` literals